### PR TITLE
Cache secret lookups, fixes #2924

### LINKED ISSUE #2924: Limit API calls to secrets backend
## What challenge are you facing?

Last month we had ~62 million requests to AWS secrets manager from Concourse. That is about 24 requests per second, every second in a 30 day period. We have 4x ATC's and 8x workers running roughly 400 containers (check and otherwise) at any point in the day, and ~500 secrets used by Concourse that are stored in secrets manager.

Without doing the math, the number of requests seem a bit excessive and represent a significant cost in AWS Secrets Manager, and added costs in Guard Duty for analysing all the secrets manager requests. For SSM I suspect it's very easy to get hit by the rate limit with this amount of requests.

## What would make this better?

The only option that come to mind is implementing a cache to avoid hitting the APIs directly with each request. Also, https://github.com/concourse/concourse/issues/2386 might already help reduce the number of API calls to the secret backend (but I have not read the implementation).

edit: Just read that global resources does not reduce the number of calls to the secret backend according to Vito https://github.com/concourse/concourse/issues/2386#issuecomment-426786169:
> Hmm I don't think this will actually help that case at all. Credential resolution is done prior to even determining whether to run the check or not, since we have to resolve the credentials to determine the resource config to check in the first place.

## Are you interested in implementing this yourself?

I'm interested, but find it hard to navigate the Concourse codebase and would need someone to point me in the right direction 🤗

### PR DESCRIPTION
**Pull request by ralekseenkov**
refactoring of secret managers + generic secret cache implementaiton

this will also serve as a good foundation for team-based credential managers RFC, as now you can introduce a composable secret manager (same way as caching and retrying is implemented), which can iterate over several underlying secret managers

```
* composable secret manager for team 1
           -> vault 1
           -> vault 2
           -> ssm
```                 

and you can add retries and caching at any layer. e.g. one global cache, or several caches per each credential manager, etc.

see commit description

### CODE REVIEW

**Comment by topherbullock** [comment]
Hey @ralekseenkov 
Question about one of the commits in this PR: https://github.com/concourse/concourse/pull/3628/commits/c0cecccb6cffe6e75cfd530f798aeb2fe8d42dbf - was there a need to add the `gobin ...` directives vs installing `counterfeiter` to the `GOPATH/$bin`(https://github.com/maxbrunsfeld/counterfeiter#installing-counterfeiter-to-gopathbin), or is this change to ensure the version of the tool is tracked `go.mod`?

**Comment by ralekseenkov** [comment]
@topherbullock my PR was created before Concourse team opted to install counterfeiter to $GOPATH/bin. I just rebased it on top of master and removed these changes

**Comment by vito** [comment]
Taking a look at this now! There are merge conflicts but I've resolved them locally (it wasn't too bad). If all is good I'll just push the merge with the conflict resolutions. :+1: 

So far I've just kicked the tires at a surface level and confirmed that credentials are cached with a locally running Vault by inspecting the audit log to see when credentials are fetched.

The default cache duration is only one minute, which means credentials will still be re-fetched roughly every resource check (which also has a one minute interval). To me that makes sense, from the standpoint that it's a conservative default that will help when there are many builds running, while still preventing a credential rotated independent of its lease duration (i.e. manually) from impacting the system for too long, as the next resource check or two will pick it up. If users need to mitigate load they can always increase the `--secret-cache-duration`.

I haven't done any testing on lease durations yet, but its interaction with `--secret-cache-duration` makes sense to me (smaller duration wins).

Thanks for starting on all this, and thanks for your patience!

File: atc/atccmd/command.go [hunk 1]
```diff
@@ -880,14 +880,20 @@ func (cmd *RunCommand) variablesFactory(logger lager.Logger) (creds.VariablesFac
...
 		}
 
 		break
 	}
-	return creds.NewRetryableVariablesFactory(variablesFactory, cmd.CredentialManagement.RetryConfig), nil
+
+	result := secretsFactory.NewSecrets()
+	result = creds.NewRetryableSecrets(result, cmd.CredentialManagement.RetryConfig)
+	if cmd.CredentialManagement.CacheConfig.Enabled {
+		result = creds.NewCachedSecrets(result, cmd.CredentialManagement.CacheConfig)
```
Comment (vito) [inline atc/atccmd/command.go]: :tada: composition!

File: atc/creds/cached_secrets.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
+package creds
+
+import (
+	"github.com/patrickmn/go-cache"
```
Comment (vito) [inline atc/creds/cached_secrets.go]: nice that there was a package for this :+1:

File: atc/creds/cached_secrets.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	// otherwise, let's make a request to the underlying secret manager
+	value, expiration, found, err := cs.secrets.Get(secretPath)
+
+	// we don't want to cache errors, let the errors be retried the next time around
+	if err != nil {
+		return nil, nil, false, err
+	}
+
+	// here we want to cache secret value, expiration, and found flag too
+	// meaning that "secret not found" responses will be cached too!
```
Comment (vito) [inline atc/creds/cached_secrets.go]: neat

File: atc/creds/secrets_factory.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+//go:generate counterfeiter . Secrets
+
+type Secrets interface {
+	// Every credential manager needs to be able to return (secret, secret_expiration_time, exists, error) based on the secret path
+	Get(string) (interface{}, *time.Time, bool, error)
+
+	// NewSecretLookupPaths returns an instance of lookup policy, which can transform pipeline ((var)) into one or more secret paths, based on team name and pipeline name
+	NewSecretLookupPaths(string, string) []SecretLookupPath
```
Comment (vito) [inline atc/creds/secrets_factory.go]: :+1: nice idea - simplifies everyone's `Get` implementation

although, should this just be called `SecretLookupPaths`? it doesn't really feel like a constructor to me

File: atc/creds/credhub/credhub.go [hunk 5]
```diff
@@ -87,14 +85,3 @@ func (c CredHubAtc) findCred(path string) (credentials.Credential, bool, error)
 
 	return cred, true, nil
 }
-
-func (c CredHubAtc) path(segments ...string) string {
-	return path.Join(append([]string{c.PathPrefix}, segments...)...)
-}
-
-func (c CredHubAtc) List() ([]template.VariableDefinition, error) {
```
Comment (vito) [inline atc/creds/credhub/credhub.go]: :+1: I like that folks don't have to implement this stub anymore

File: atc/creds/secretsmanager/secretsmanager.go [hunk 6]
```diff
@@ -3,69 +3,60 @@ package secretsmanager
...
-			return nil, false, err
-		}
-		if found {
-			return value, true, nil
-		}
+// Get retrieves the value and expiration of an individual secret
+func (s *SecretsManager) Get(secretPath string) (interface{}, *time.Time, bool, error) {
+	value, expiration, found, err := s.getSecretById(secretPath)
+	if err != nil {
+		s.log.Error("unable to retrieve aws secret", err, lager.Data{
```
Comment (vito) [inline atc/creds/secretsmanager/secretsmanager.go]: this should be conventional formatting:

```suggestion
		s.log.Error("failed-to-fetch-aws-secret", err, lager.Data{
```

**Comment by ralekseenkov** [comment]
Sounds good. I was trying to be careful, decoupling variable lookup logic (i.e. NewSecretLookupPaths) from secret retrieval. Hopefully nothing got broken as a result of this.

But there are cases like Kubernetes credential manager - it doesn't seem to have any tests and it's currently missing docs, see https://github.com/concourse/concourse/issues/2003. That makes it challenging to verify

**Comment by ralekseenkov** [comment]
thanks :)

**Comment by samgurtman** [comment]
I hate to do this, but would anyone have any guidance on back porting this to 5.1.0? I have a big of egg on my face after moving our credentials to SSM. I guess I can set the default resource check to much longer, but this seems like a way better solution.

**Comment by ralekseenkov** [comment]
@samgurtman it might be easier to wait for 5.2.0 because maintaining your own forks is never fun. if you really want to backport, ping me on Discord and I can try to help

**Comment by cirocosta** [comment]
Hey @ralekseenkov, I noticed that the kubernetes cred mgmt changed a bit, breaking the tests we have in place for such integration:

https://github.com/concourse/concourse/blob/37dd02b20741a9c90b362f3f4edeb918a674caa9/topgun/k8s/kubernetes_creds_mgmt_test.go#L73-L120

Seems related to the new idea of having the `:` separator here:

https://github.com/concourse/concourse/blob/7b846eeac9a5ca232d521f25b2c0c410e72e02de/atc/creds/kubernetes/kubernetes.go#L33-L35

Was there a specific reason for going this way?

Thanks!

**Comment by cirocosta** [comment]
Oh, here's an example failure:

```
 running: kubectl create secret generic foo --namespace=topgun-k8s-cm-664743-1-main --from-literal=value=bar
secret/foo created
running: kubectl create secret generic caz --namespace=topgun-k8s-cm-664743-1-main --from-literal=baz=zaz
secret/caz created

fly -t concourse-topgun-k8s-1 set-pipeline -n -c ../pipelines/minimal-credential-management.yml -p pipeline
jobs:
  job unit has been added:
+ name: unit
+ plan:
+ - task: run
+   config:
+     platform: linux
+     image_resource:
+       type: registry-image
+       source:
+         repository: busybox
+     run:
+       path: /bin/sh
+       args:
+       - -c
+       - |
+         echo plain-secret=((foo))
+         echo nested-secret=((caz.baz))

...

failed to interpolate task config: 
  Finding variable 'foo': unable to split kubernetes secret path 
  into [namespace]:[secret]: topgun-k8s-cm-664743-1-/main/pipeline/foo
```

**Comment by ralekseenkov** [comment]
@cirocosta it should be backwards compatible. I'll look into this today

**Pull request merged**